code/modules/shuttle/supply.dm

```diff
@@ -108,131 +115,511 @@
-						service_credits += S.rarity / 2 // That's right, no bonus for potency. Send a crappy sample first to "show improvement" later
-						credits_to_deposit += S.rarity / 2
-
-				if(istype(thing, /obj/item/organ/internal/alien))
-					var/obj/item/organ/internal/alien/organ = thing
-					credits_to_deposit += organ.cargo_profit
-		qdel(MA)
-		SSeconomy.sold_atoms += "."
-
-	if(plasma_count > 0)
-		var/credits_from_plasma = plasma_count * SSeconomy.credits_per_plasma
-		msg += "<span class='good'>+[credits_from_plasma]</span>: Received [plasma_count] unit(s) of exotic material.<br>"
-		credits_to_deposit += credits_from_plasma
-
-	if(intel_count > 0)
-		var/credits_from_intel = intel_count * SSeconomy.credits_per_intel
-		msg += "<span class='good'>+[credits_from_intel]</span>: Received [intel_count] article(s) of enemy intelligence.<br>"
-		credits_to_deposit += credits_from_intel
-
-	if(crate_count > 0)
-		msg += "<span class='good'>+[total_crate_value]</span>: Received [crate_count] crate(s).<br>"
-		credits_to_deposit += total_crate_value
+	var/list/credit_changes = list()
+	for(var/datum/economy/line_item/item in manifest.line_items)
+		if(!credit_changes[item.account])
+			credit_changes[item.account] = 0
+		credit_changes[item.account] += item.credits
+
+		if(item.credits > 0)
+			msg += "<span class='good'>[item.account.account_name] +[item.credits]</span>: [item.reason]<br>"
+		else
+			msg += "<span class='bad'>[item.account.account_name] [item.credits]</span>: [item.reason]<br>"
+
+	for(var/account in credit_changes)
+		if(credit_changes[account] > 0)
+			GLOB.station_money_database.credit_account(account, credit_changes[account], "Supply Shuttle Exports Payment", "Central Command Supply Master", supress_log = FALSE)
+		else
+			GLOB.station_money_database.charge_account(account, -credit_changes[account], "Supply Shuttle Fine", "Central Command Supply Master", allow_overdraft = TRUE, supress_log = FALSE)
 
 	SSeconomy.centcom_message += "[msg]<hr>"
-	if(credits_to_deposit > 0)
-		GLOB.station_money_database.credit_account(SSeconomy.cargo_account, credits_to_deposit, "Supply Shuttle Exports Payment", "Central Command Supply Master", supress_log = FALSE)
-		if(research_credits)
-			GLOB.station_money_database.credit_account(GLOB.station_money_database.get_account_by_department(DEPARTMENT_SCIENCE), research_credits, "Supply Shuttle Exports Payment", "Central Command Supply Master", supress_log = FALSE)
-		if(service_credits)
-			GLOB.station_money_database.credit_account(GLOB.station_money_database.get_account_by_department(DEPARTMENT_SERVICE), service_credits, "Supply Shuttle Exports Payment", "Central Command Supply Master", supress_log = FALSE)
+	manifest = new
+
+
+// Convenience object that registers itself with the supply shuttle and provides
+// methods for you to override.
+/datum/economy/simple_seller
+
+/datum/economy/simple_seller/proc/register(obj/docking_port/mobile/supply/S)
```

Comment:
I'm a little confused about the use of signals in this. As I understand it, you have a set of simple seller datums. Each one provides specific rules on what to do when an atom is encountered on the shuttle, and what to add to the manifest when encountered. Every supply shuttle port gets these sellers attached to it on initialize. All of the signals are keyed one-to-one with the functions on the seller with RegisterSignal. This all seems fairly self contained. Why are we sending e.g. a BEGIN_SELL signal to all of the sellers registered with us instead of just iterating over them and calling `begin_sell()`?

Reply:
`simple_seller` is just that, simple. If that were the only implementation we'd ever have, then doing a for loop would work fine, though you'd end up with a bunch of them scattered around.

#24599 is going to be adding goals that use this system to hook in and consume runtime-defined items, which doesn't really fit the pattern of simple_seller. You could shoehorn it in if you tried, but at that point, you're just re-implementing signals, so why not use them in the first place?

Reply:
So effectively, there will be goals introduced that would also listen in on these signals?